File: lib/features/profile/domain/logros.dart
/// Catálogo de logros (Fase 9 — capa domain, sin Flutter ni Firebase).
///
/// Espejo de la metadata de `functions/src/logros.ts`: aquí están el nombre, el
/// emoji y la descripción para la **galería** de la UI; allá están las
/// condiciones que la Cloud Function evalúa server-side (anti-trampa). Los `id`
/// deben coincidir exactamente en ambos lados. Al añadir/cambiar un logro,
/// actualizar los dos archivos.
library;

/// Definición de un logro para mostrarlo en la galería.
class Logro {
  const Logro({
    required this.id,
    required this.emoji,
    required this.nombre,
    required this.descripcion,
  });

  final String id;
  final String emoji;
  final String nombre;

  /// Cómo se desbloquea (texto para la UI).
  final String descripcion;
}

/// Todos los logros, en el orden en que se muestran en la galería.
const List<Logro> catalogoLogros = [
  Logro(
    id: 'primera',
    emoji: '🎉',
    nombre: 'Primera victoria',
    descripcion: 'Gana tu primera mano.',
  ),
  Logro(
    id: 'bj',
    emoji: '🃏',
    nombre: '¡Blackjack!',
    descripcion: 'Consigue un blackjack.',
  ),
  Logro(
    id: 'racha3',
    emoji: '🔥',
    nombre: '3 seguidas',
    descripcion: 'Gana 3 rondas consecutivas.',
  ),
  Logro(
    id: 'racha5',
    emoji: '🔥🔥',
    nombre: '5 seguidas',
    descripcion: 'Gana 5 rondas consecutivas.',
  ),
  Logro(
    id: 'veterano',
    emoji: '🎖️',
    nombre: 'Veterano',
    descripcion: 'Juega 50 manos.',
  ),
  Logro(
    id: 'centenario',
    emoji: '💯',
    nombre: 'Centenario',
    descripcion: 'Juega 100 manos.',
  ),
  Logro(
    id: 'granGanancia',
    emoji: '💰',
    nombre: 'Gran ganancia',
    descripcion: 'Gana 500 o más en una sola ronda.',
  ),
  Logro(
    id: 'ricachon',
    emoji: '💎',
    nombre: 'Ricachón',
    descripcion: 'Alcanza un saldo de 5000.',
  ),
];

/// Busca un logro por su [id]; `null` si no está en el catálogo (p. ej. un id
/// nuevo escrito por una versión más reciente de la Function).
Logro? logroPorId(String id) {
  for (final l in catalogoLogros) {
    if (l.id == id) return l;
  }
  return null;
}

File: functions/src/logros.ts
// ─────────────────────────────────────────────────────────────────────────────
// Evaluación de logros server-side (Fase 9). Anti-trampa: la Function decide
// qué logros se desbloquean al resolver una ronda y los persiste en
// users/{uid}.logros (array de IDs). El cliente solo los lee y muestra.
//
// La METADATA de los logros (nombre, emoji, descripción) vive en el espejo Dart
// lib/features/profile/domain/logros.dart para la galería de la UI. Aquí solo
// están los IDs y sus condiciones. Al añadir/cambiar un logro, actualizar ambos.
// ─────────────────────────────────────────────────────────────────────────────

import { EstadisticasJugador } from './blackjack';

/** Condición de un logro a partir de las stats acumuladas y el saldo actual. */
interface DefinicionLogro {
  id: string;
  /** Nombre legible (para el texto del push); espejo de `logros.dart`. */
  nombre: string;
  cond: (stats: EstadisticasJugador, balance: number) => boolean;
}

const LOGROS: DefinicionLogro[] = [
  { id: 'primera', nombre: '🎉 Primera victoria', cond: (s) => s.ganadas >= 1 },
  { id: 'bj', nombre: '🃏 ¡Blackjack!', cond: (s) => s.blackjacks >= 1 },
  { id: 'racha3', nombre: '🔥 3 seguidas', cond: (s) => s.mejorRacha >= 3 },
  { id: 'racha5', nombre: '🔥🔥 5 seguidas', cond: (s) => s.mejorRacha >= 5 },
  { id: 'veterano', nombre: '🎖️ Veterano', cond: (s) => s.manosJugadas >= 50 },
  { id: 'centenario', nombre: '💯 Centenario', cond: (s) => s.manosJugadas >= 100 },
  { id: 'granGanancia', nombre: '💰 Gran ganancia', cond: (s) => s.mayorGanancia >= 500 },
  { id: 'ricachon', nombre: '💎 Ricachón', cond: (_s, balance) => balance >= 5000 },
];

/** Nombre legible de un logro por su id; el propio id si no se encuentra. */
export function nombreLogro(id: string): string {
  return LOGROS.find((l) => l.id === id)?.nombre ?? id;
}

/**
 * Devuelve los IDs de logros recién desbloqueados (no presentes en
 * `yaDesbloqueados`) según las stats y el saldo. Función pura.
 */
export function evaluarLogros(
  stats: EstadisticasJugador,
  balance: number,
  yaDesbloqueados: string[],
): string[] {
  const previos = new Set(yaDesbloqueados);
  const nuevos: string[] = [];
  for (const l of LOGROS) {
    if (!previos.has(l.id) && l.cond(stats, balance)) {
      nuevos.push(l.id);
    }
  }
  return nuevos;
}
